File: downloader/manager.py
from typing import List
from downloader.downloader_factory import DownloaderFactory
from downloader.status_manager import StatusManager


class DownloadManager:
    def __init__(self, factory: DownloaderFactory):
        self.status_manager = StatusManager()
        self.file_downloader = factory.create_file_downloader(
            save_dir='files', manager=self.status_manager)
        self.youtube_downloader = factory.create_youtube_downloader(
            save_dir='youtube_videos', manager=self.status_manager)
        self.instagram_downloader = factory.create_instagram_downloader(
            save_dir='instagram_posts', manager=self.status_manager)
        self.file_urls: List[str] = []
        self.youtube_urls: List[str] = []
        self.instagram_urls: List[str] = []

    def add_file_url(self, url: str) -> None:
        self.file_urls.append(url)
        self.status_manager.update_status(url, 'Pending')

    def add_youtube_url(self, url: str) -> None:
        self.youtube_urls.append(url)
        self.status_manager.update_status(url, 'Pending')

    def add_instagram_url(self, url: str) -> None:
        self.instagram_urls.append(url)
        self.status_manager.update_status(url, 'Pending')

    def remove_download(self, url: str) -> None:
        if url in self.file_urls:
            self.file_urls.remove(url)
        if url in self.youtube_urls:
            self.youtube_urls.remove(url)
        if url in self.instagram_urls:
            self.instagram_urls.remove(url)
        self.status_manager.remove_status(url)

    def update_status(self, url: str, name: str, status: str) -> None:
        self.status_manager.update_status(url, f"{name[:10]}: {status}")

    def download_all(self) -> None:
        if self.file_urls:
            self.file_downloader.download_files(self.file_urls)
        if self.youtube_urls:
            self.youtube_downloader.download_videos(self.youtube_urls)
        if self.instagram_urls:
            self.instagram_downloader.download_posts(self.instagram_urls)

File: src/main/downloader/managers/manager.ts
import { DownloaderFactory } from '../downloaders/downloader_factory'
import { FileDownloader } from '../downloaders/files'
import { YouTubeDownloader } from '../downloaders/youtube'
import { StatusManager } from './status_manager'

type UrlList = string[]

class DownloadManager {
  statusManager: StatusManager
  private fileDownloader: FileDownloader
  private youtubeDownloader: YouTubeDownloader
  private fileUrls: UrlList
  private youtubeUrls: UrlList

  constructor(factory: DownloaderFactory) {
    this.statusManager = new StatusManager()
    this.fileDownloader = factory.createFileDownloader('files', this.statusManager)
    this.youtubeDownloader = factory.createYouTubeDownloader('youtube_videos', this.statusManager)
    this.fileUrls = []
    this.youtubeUrls = []
  }

  private urlType(url: string): 'youtube' | 'file' {
    if (url.includes('youtube') || url.includes('youtu.be')) {
      return 'youtube'
    }

    return 'file'
  }

  addUrl(url: string): void {
    const urlType = this.urlType(url)

    switch (urlType) {
      case 'youtube':
        this.youtubeUrls.push(url)
        this.statusManager.updateStatus(url, 'Pending')

        break

      default:
        this.fileUrls.push(url)
        this.statusManager.updateStatus(url, 'Pending')
        break
    }
  }

  addUrls(urls: string[]): void {
    urls.forEach((u) => this.addUrl(u))
  }

  removeDownload(url: string): void {
    this.fileUrls = this.fileUrls.filter((fileUrl) => fileUrl !== url)
    this.youtubeUrls = this.youtubeUrls.filter((youtubeUrl) => youtubeUrl !== url)
    this.statusManager.removeStatus(url)
  }

  updateStatus(url: string, status: string): void {
    this.statusManager.updateStatus(url, status)
  }

  async downloadAll(): Promise<void> {
    if (this.fileUrls.length > 0) {
      await this.fileDownloader.downloadFiles(this.fileUrls)
    }
    if (this.youtubeUrls.length > 0) {
      await this.youtubeDownloader.downloadVideos(this.youtubeUrls)
    }
  }
}

export { DownloadManager }
